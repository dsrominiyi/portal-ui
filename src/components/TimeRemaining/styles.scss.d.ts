export const circularProgressBar: string;
export const green: string;
export const red: string;
export const amber: string;
export const circularProgressBarTrail: string;
export const circularProgressBarText: string;
export const circularProgressBarBackground: string;
export const circularProgressBarInverted: string;
export const circularProgressBarPath: string;
