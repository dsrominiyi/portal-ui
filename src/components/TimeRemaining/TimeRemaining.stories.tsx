import React from 'react';
import { storiesOf } from '@storybook/react';
import TimeRemaining, { ConfigPercentages } from '.';

const stories = storiesOf('TimeRemaining', module);

const start = '2019-01-01';
const end = '2019-10-01';
const current = '2019-05-01';

const config: ConfigPercentages = {
  green: 0.7,
  amber: 0.1
};

const render = (
  startDate: string = start,
  endDate: string = end,
  currentDate?: string,
  configPercentages?: ConfigPercentages
) => () => (
  <TimeRemaining
    startDate={startDate}
    endDate={endDate}
    currentDate={currentDate}
    configPercentages={configPercentages}
  />
);

export const storyRenders = {
  green: render(start, end, '2019-01-02'),
  amber: render('2019-10-01', '2019-10-20', '2019-10-15'),
  red: render('2019-10-01', '2019-10-10', '2019-10-09'),
  lapsed: render(start, end, '2019-10-05'),
  currentDate: render('2019-10-01', '2020-10-01'),
  customConfig: render(start, end, current, config),
  oneDayRemaining: render(start, end, '2019-10-01')
};

stories.add('Green', storyRenders.green);
stories.add('Amber', storyRenders.amber);
stories.add('Red', storyRenders.red);
stories.add('Lapsed', storyRenders.lapsed);
stories.add('Current date', storyRenders.currentDate);
stories.add('Custom config', storyRenders.customConfig);
stories.add('One day remaining', storyRenders.oneDayRemaining);
