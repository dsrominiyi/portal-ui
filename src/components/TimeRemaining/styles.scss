.circular-progress-bar {
  width: 100%;
  max-width: 250px;
  vertical-align: middle;
}

.circular-progress-bar %circular-progress-bar-path {
  stroke-linecap: butt;
  transition: stroke-dashoffset 0.5s ease 0s;
}

.green {
  @extend %circular-progress-bar-path;

  stroke: #92d400;
}

.red {
  @extend %circular-progress-bar-path;

  stroke: #f25544;
}

.amber {
  @extend %circular-progress-bar-path;

  stroke: #ffbf00;
}

.circular-progress-bar .circular-progress-bar-trail {
  stroke: #d6d6d6;
  stroke-linecap: round;
}

.circular-progress-bar .circular-progress-bar-text {
  fill: #666;
  font-size: 20px;
  font-family: 'BrandonText-Regular', Arial, Helvetica, sans-serif;
  dominant-baseline: middle;
  text-anchor: middle;
}

.circular-progress-bar .circular-progress-bar-background {
  fill: #d6d6d6;
}

.circular-progress-bar.circular-progress-bar-inverted .circular-progress-bar-background {
  fill: #3e98c7;
}

.circular-progress-bar.circular-progress-bar-inverted .circular-progress-bar-text {
  fill: #fff;
}

.circular-progress-bar.circular-progress-bar-inverted .circular-progress-bar-path {
  stroke: #fff;
}

.circular-progress-bar.circular-progress-bar-inverted .circular-progress-bar-trail {
  stroke: transparent;
}
