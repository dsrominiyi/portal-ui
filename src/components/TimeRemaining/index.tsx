import React, { FunctionComponent } from 'react';
import { CircularProgressbar } from 'react-circular-progressbar';
import { utc as moment, duration } from 'moment';
import styles from './styles.scss';

export interface ConfigPercentages {
  green: number;
  amber: number;
}

interface Props {
  startDate: string;
  endDate: string;
  currentDate?: string;
  configPercentages?: ConfigPercentages;
}

export const TimeRemaining: FunctionComponent<Props> = ({
  startDate,
  endDate,
  currentDate,
  configPercentages
}: Props) => {
  const startDateUtc = moment(startDate, 'YYYY-MM-DD');
  const endDateUtc = moment(endDate, 'YYYY-MM-DD');
  const currentDateUtc = currentDate ? moment(currentDate, 'YYYY-MM-DD') : moment();

  const policyLength = Math.floor(duration(endDateUtc.diff(startDateUtc)).asDays() + 1);
  const daysRemaining = Math.ceil(duration(endDateUtc.diff(currentDateUtc)).asDays() + 1);

  const displayText =
    daysRemaining > 0 ? `${daysRemaining} day${daysRemaining === 1 ? '' : 's'}` : 'Lapsed';
  const percentageRemaining = daysRemaining / policyLength;
  const greenFloat = configPercentages ? configPercentages.green : 0.331;
  const amberFloat = configPercentages ? configPercentages.amber : 0.2501;
  const pathColour =
    percentageRemaining > greenFloat ? 'green' : percentageRemaining > amberFloat ? 'amber' : 'red';

  return (
    <CircularProgressbar
      value={daysRemaining}
      minValue={0}
      maxValue={policyLength}
      text={displayText}
      counterClockwise={false}
      strokeWidth={5}
      classes={{
        root: styles.circularProgressBar,
        trail: styles.circularProgressBarTrail,
        path: styles[pathColour],
        text: styles.circularProgressBarText,
        background: styles.circularProgressBarBackground
      }}
    />
  );
};

TimeRemaining.displayName = 'TimeRemaining';
export default TimeRemaining;
