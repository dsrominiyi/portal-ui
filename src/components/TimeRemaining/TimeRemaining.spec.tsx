import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { utc as moment } from 'moment';

import { storyRenders } from './TimeRemaining.stories';

import testStorySnapshots from '../../helpers/test/testStorySnapshots';

let component: ShallowWrapper;

describe('TimeRemaining', () => {
  test('component story renders match snapshots', () => {
    testStorySnapshots(storyRenders, ['currentDate']);
  });

  it('should match the snapshot if current date is not passed in', () => {
    const props = {
      startDate: '2019-01-01',
      endDate: '2019-10-01'
    };

    const startDate = moment(props.startDate, 'YYYY-MM-DD');
    const endDate = moment(props.endDate, 'YYYY-MM-DD');
    const currentDate = moment('2019-05-01', 'YYYY-MM-DD');

    const momentMock = jest.fn();
    const utcMock = jest.fn();
    utcMock
      .mockImplementationOnce(() => startDate)
      .mockImplementationOnce(() => endDate)
      .mockImplementationOnce(() => currentDate);

    const durationMock = { asDays: jest.fn() };
    (momentMock as any).duration = jest.fn(() => durationMock);
    (momentMock as any).utc = utcMock;
    durationMock.asDays.mockImplementationOnce(() => 273).mockImplementationOnce(() => 153);

    jest.doMock('moment', () => momentMock);
    jest.resetModules();
    const TimeRemaining = require('.').default;

    component = shallow(<TimeRemaining {...props} />);

    expect(component).toMatchSnapshot();
  });
});
