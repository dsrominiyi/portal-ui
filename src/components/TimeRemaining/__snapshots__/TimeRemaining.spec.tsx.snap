// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeRemaining component story renders match snapshots: amber 1`] = `
<CircularProgressbar
  background={false}
  backgroundPadding={0}
  circleRatio={1}
  className=""
  classes={
    Object {
      "background": "circularProgressBarBackground",
      "path": "amber",
      "root": "circularProgressBar",
      "text": "circularProgressBarText",
      "trail": "circularProgressBarTrail",
    }
  }
  counterClockwise={false}
  maxValue={20}
  minValue={0}
  strokeWidth={5}
  styles={
    Object {
      "background": Object {},
      "path": Object {},
      "root": Object {},
      "text": Object {},
      "trail": Object {},
    }
  }
  text="6 days"
  value={6}
/>
`;

exports[`TimeRemaining component story renders match snapshots: customConfig 1`] = `
<CircularProgressbar
  background={false}
  backgroundPadding={0}
  circleRatio={1}
  className=""
  classes={
    Object {
      "background": "circularProgressBarBackground",
      "path": "amber",
      "root": "circularProgressBar",
      "text": "circularProgressBarText",
      "trail": "circularProgressBarTrail",
    }
  }
  counterClockwise={false}
  maxValue={274}
  minValue={0}
  strokeWidth={5}
  styles={
    Object {
      "background": Object {},
      "path": Object {},
      "root": Object {},
      "text": Object {},
      "trail": Object {},
    }
  }
  text="154 days"
  value={154}
/>
`;

exports[`TimeRemaining component story renders match snapshots: green 1`] = `
<CircularProgressbar
  background={false}
  backgroundPadding={0}
  circleRatio={1}
  className=""
  classes={
    Object {
      "background": "circularProgressBarBackground",
      "path": "green",
      "root": "circularProgressBar",
      "text": "circularProgressBarText",
      "trail": "circularProgressBarTrail",
    }
  }
  counterClockwise={false}
  maxValue={274}
  minValue={0}
  strokeWidth={5}
  styles={
    Object {
      "background": Object {},
      "path": Object {},
      "root": Object {},
      "text": Object {},
      "trail": Object {},
    }
  }
  text="273 days"
  value={273}
/>
`;

exports[`TimeRemaining component story renders match snapshots: lapsed 1`] = `
<CircularProgressbar
  background={false}
  backgroundPadding={0}
  circleRatio={1}
  className=""
  classes={
    Object {
      "background": "circularProgressBarBackground",
      "path": "red",
      "root": "circularProgressBar",
      "text": "circularProgressBarText",
      "trail": "circularProgressBarTrail",
    }
  }
  counterClockwise={false}
  maxValue={274}
  minValue={0}
  strokeWidth={5}
  styles={
    Object {
      "background": Object {},
      "path": Object {},
      "root": Object {},
      "text": Object {},
      "trail": Object {},
    }
  }
  text="Lapsed"
  value={-3}
/>
`;

exports[`TimeRemaining component story renders match snapshots: oneDayRemaining 1`] = `
<CircularProgressbar
  background={false}
  backgroundPadding={0}
  circleRatio={1}
  className=""
  classes={
    Object {
      "background": "circularProgressBarBackground",
      "path": "red",
      "root": "circularProgressBar",
      "text": "circularProgressBarText",
      "trail": "circularProgressBarTrail",
    }
  }
  counterClockwise={false}
  maxValue={274}
  minValue={0}
  strokeWidth={5}
  styles={
    Object {
      "background": Object {},
      "path": Object {},
      "root": Object {},
      "text": Object {},
      "trail": Object {},
    }
  }
  text="1 day"
  value={1}
/>
`;

exports[`TimeRemaining component story renders match snapshots: red 1`] = `
<CircularProgressbar
  background={false}
  backgroundPadding={0}
  circleRatio={1}
  className=""
  classes={
    Object {
      "background": "circularProgressBarBackground",
      "path": "red",
      "root": "circularProgressBar",
      "text": "circularProgressBarText",
      "trail": "circularProgressBarTrail",
    }
  }
  counterClockwise={false}
  maxValue={10}
  minValue={0}
  strokeWidth={5}
  styles={
    Object {
      "background": Object {},
      "path": Object {},
      "root": Object {},
      "text": Object {},
      "trail": Object {},
    }
  }
  text="2 days"
  value={2}
/>
`;

exports[`TimeRemaining should match the snapshot if current date is not passed in 1`] = `
<CircularProgressbar
  background={false}
  backgroundPadding={0}
  circleRatio={1}
  className=""
  classes={
    Object {
      "background": "circularProgressBarBackground",
      "path": "green",
      "root": "circularProgressBar",
      "text": "circularProgressBarText",
      "trail": "circularProgressBarTrail",
    }
  }
  counterClockwise={false}
  maxValue={274}
  minValue={0}
  strokeWidth={5}
  styles={
    Object {
      "background": Object {},
      "path": Object {},
      "root": Object {},
      "text": Object {},
      "trail": Object {},
    }
  }
  text="154 days"
  value={154}
/>
`;
